### ISSUE TITLE
Introduce nix compatibility modes

### ISSUE THREAD

**Issue by LnL7** [NixOS/nix#3663]
Nix releases regularly introduce new builtins or other language features, but currently there's not really a way to declare what version of nix the expression expects. In nixpkgs this is currently done with `builtins.nixVersion` but exposing that is kind of impure and enforcing that it actually stays compatible is very tricky. Using an old version of nix for CI is one way of ensuring compatibility but that's not really a desirable approach.
Not having this also means that changing behaviour that's not backwards compatible (eg. deprecating sha256 in favour of hash) would be a major braking change that would be hard to transition.

Flakes, unlike plain nix expressions, provide a nice place to define this requirement. Either by declaring the minimum version similar to minvers.nix in nixpkgs or perhaps even specific features in some cases (plugins?, etc.). Flake inputs can also be validated that they don't require a newer version so updating doesn't accidentally break the compatibility.

Secondly introducing evaluation modes that can be enabled (automatically?) which disable builtins and other functionality added in later releases. To make it easy to ensure compatibility when running a newer version of nix as well as providing a nice indication of when the version requirement should be bumped for the nixos/* flakes.

Some examples of this in nixpkgs:
- placeholders, mostly mitigated by using nix1 in ofborg during the compatibility window.
- SRI hashes, hasn't been a big problem since it didn't happen to core packages until just recently but could have caused much bigger issues for users.

**Comment by edolstra** [comment]
The flake edition field was intended to serve this purpose. However I removed it. Now the idea is that we'll change the `flake.nix` file name if the language changes (e.g. `flake-v2.nix`).

**Comment by nrdxp** [comment]
I was going to open an issue to suggest specifying the exact Nix derivation used to evaluate a flake (to solve exactly the same problem), but this seems like a fitting place to put it instead. Something like:
```nix
# flake.nix
{
  inputs.nix.url = "github:NixOS/nix/someArbitraryRef";
  nixVersion = self.inputs.nix;
}
```
The `nix` binary could do something similar to how it resolves the current system, to pull the proper system out of the nix flake. The downside is that Nix would then have to go download (or build) that copy of Nix in order to continue evaluation.

**Comment by ajs124** [comment]
> The compatibility mode itself can be mostly separate from flakes, except for optionally using the version specified by the flake as default instead of latest.

Agreed. I don't use flakes at all and this is still a quite relevant issue to me.
